### PR TITLE
RNNs with transferable hidden state

### PR DESCRIPTION
**Pull request by farizrahman4u**
* Hidden state is not reset after every batch
* Can transfer hidden state from one RNN to another

### CODE REVIEW

**Comment by farizrahman4u** [comment]
**API**:

* Set the `stateful` argument to `True`
* When the `stateful` argument is `True`, one of the following arguments should also be provided: `hidden_state`, `batch_size`, `weights`.
* `hidden_state`:  Used to initialize with hidden state of another RNN.
* `batch_size`: Used to initialize the hidden state with zeroes.
* `weights` : The `weights` argument can be used to set the weights and hidden state together.
For example:  `RNN1.set_weights(RNN2.get_weights())` will copy both weights and hidden states. If either `RNN1 or `RNN2` are stateless, only the weights will be copied.

**Comment by fchollet** [comment]
Looks like Travis is failing again. We will probably have to update the Travis config to fix that. I'll look at it later today / tomorrow.

File: keras/layers/recurrent.py [hunk 1]
```diff
@@ -78,11 +128,27 @@ def build(self):
...
+                    raise Exception("Hidden state not provided in weights")
+            else:
+                raise Exception("One of the following arguments must be provided for stateful RNNs: hidden_state, batch_size, weights")
+        self.state = [self.h]
         if self.initial_weights is not None:
-            self.set_weights(self.initial_weights)
+            self.set_weights(self.initial_weights[:nw])
             del self.initial_weights
 
+
```
Comment (EderSantana) [inline keras/layers/recurrent.py]: the next _step function uses `return self.activation(x_t + mask_tm1 * T.dot(h_tm1, u))` which makes statefulness completly useless when using different length sequences. It should instead return
```python
h_t = self.activation( x_t + T.dot(h_tm1, u))
return mask * h_t + (1 - mask) * h_tm1
```
I actually have a question. What is faster inside a scan loop?
* `mask * h_t + (1 - mask) * h_tm1` or
*  `T.switch(T.eq(mask, 1), h_t, h_tm1)`

@nouiz, if you have time, could you please say your opinion? Will both have the same gradient optimized gradient?

**Comment by elanmart** [comment]
Shouldn't we have a special argument to `Model.fit` method to actually use statefullness in keras? 

Without smart re-shuffling, the net will see row number `i` of the data, update its hidden state, and then proceed to use this hidden state at row `i + batch_size`, which is not what we want.

**Comment by EderSantana** [comment]
@elanmart that will still work if the user organized the batches before hand to account for that behavior. Which is actually too sensitive to choices of batch size and stuff. For sure this needs a careful implementation or high user awareness.

**Comment by elanmart** [comment]
> Which is actually too sensitive to choices of batch size and stuff. For sure this needs a careful implementation or high user awareness.

Agree, it's not clear to me, however, how this should be implemented: it's not only about reshuffling the data, it's also about reseting the hidden states: the last row of layer's `hidden_state` will reach the end of the data in one epoch, while the first row will take `batch_size * nb_epoch` to do it.

**Comment by EderSantana** [comment]
the number of epochs that it takes before resetting the hidden state is usually an open parameter for the user. I saw people a number like 4 in text modeling paper. It will also depend on the average length of your phrases. If they usually spread across 4 batches, that should be how long you wait before resetting.

**Comment by fchollet** [comment]
Test are failing. 

Rather than waiting for tests to fail on Travis, I recommend running them locally first. Install [py.test](http://pytest.org/latest/), run `py.test tests/` from the Keras folder.

**Comment by farizrahman4u** [comment]
Current code is functional. Now working on statefulness for sequences of variable lengths

**Comment by EderSantana** [comment]
nice @farizrahman4u! I PRed a suggestion of RNN reset state callback check it out in your patch-14

**Comment by farizrahman4u** [comment]
@EderSantana Thanks for the input, really helped me. I can not merge it as it is, because the code was intended for Seya's stateful GRUs, but my PR contains stateful LSTMs whose state is defined by two matrices. (h and c). Also I have to take into account multiple stateful RNNs in a single model.
Something like:
```python
class ResetRNNState(Callback):
    """
    This is supposed to be used with stateful RNNs
    
    Parameters:
    ===========
    rnns: stateful RNNs
    func: a function that returns true when the states should be reset.
    states: initial state to which we reset the model when `func` is True.
    """
    def __init__(self, rnns, func, states=None):
        for rnn in rnns:
            if not hasattr(rnn, 'state'):
                raise Exception("Not stateful RNN")
        self.rnns = rnns
        self.states = None
        self.func = func

    def on_batch_end(self, batch, logs={}):
        if self.func(batch, logs):
            if self.states:
                for i in range(len(self.rnns)):
                    self.rnns[i].set_hidden_state(states[i])
            else:
                for rnn in self.rnns:
                    rnn.reset_hidden_state()

```

Can you resubmit a PR with the above code? (So that I dont take credit for your idea!)

**Comment by farizrahman4u** [comment]
@EderSantana Also i need some help on the step equations for making stateful RNNs work with variable length sequences.

**Comment by EderSantana** [comment]
@farizrahman4u all right I'll check it out and make the proper updates to the code

**Pull request closed without merging**